### PR TITLE
feat(mason_cli): support bundling git and hosted bricks

### PR DESCRIPTION
**Pull request by renancaraujo**
## Status

**READY**

## Description

This allows people to run `mason bundle` on git or brickhub-hosted brick templates. 

- Add an option to explicitly specify the source:

Registry: 

```bash
mason bundle some_brick --source=registry
mason bundle some_brick -s registry
```

Git: 

```bash
mason bundle https://github.com/felangel/mason --source=git
mason bundle https://github.com/felangel/mason -s git --git-path bricks/greeting
```

The option "source" was inspired by darts `dart pub global activate --source`.

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/mason_cli/lib/src/commands/bundle.dart [hunk 1]
```diff
@@ -46,15 +64,45 @@ class BundleCommand extends MasonCommand {
...
     }
-    final brick = Directory(results.rest.first);
-    if (!brick.existsSync()) {
-      throw BrickNotFoundException(brick.path);
+
+    final String brickPath;
+    if (brick.location.isLocal) {
+      brickPath = brick.location.path!;
+    } else {
+      final cachedBrick = await globalBricksJson.add(brick);
```
Comment (renancaraujo) [inline packages/mason_cli/lib/src/commands/bundle.dart:101]: Save on global cache if it is a remote brick template.

File: packages/mason_cli/lib/src/commands/bundle.dart [hunk 2]
```diff
@@ -46,15 +64,45 @@ class BundleCommand extends MasonCommand {
...
+            path: results['git-path'] as String?,
+            ref: results['git-ref'] as String?,
+          ),
+        ),
+      );
+    } else if (source == 'hosted') {
+      if (results.rest.isEmpty) {
+        usageException('A brick name must be provided');
+      }
+      brick = Brick(name: rest, location: const BrickLocation(version: 'any'));
```
Comment (felangel) [inline packages/mason_cli/lib/src/commands/bundle.dart:89]: should users be able to specify a specific version to bundle?

File: packages/mason_cli/test/fixtures/make/flavors/main_development.dart [hunk 3]
```diff
@@ -1,3 +1,3 @@
 void main() {
   print('Running in development mode...');
-}
\ No newline at end of file
+}
```
Comment (felangel) [inline packages/mason_cli/test/fixtures/make/flavors/main_development.dart:3]: can we revert these changes?

File: packages/mason_cli/test/fixtures/make/flavors/main_production.dart [hunk 4]
```diff
@@ -1,3 +1,3 @@
 void main() {
   print('Running in production mode...');
-}
\ No newline at end of file
+}
```
Comment (felangel) [inline packages/mason_cli/test/fixtures/make/flavors/main_production.dart:3]: can we revert this as well?

File: packages/mason/lib/src/bundler.dart [hunk 5]
```diff
@@ -42,22 +42,25 @@ void unpackBundle(MasonBundle bundle, Directory target) {
   if (license != null) _unbundleFile(license, target.path);
 }
 
-/// Generates a [MasonBundle] from the provided [brick] directory.
-MasonBundle createBundle(Directory brick) {
-  final brickYamlFile = File(path.join(brick.path, BrickYaml.file));
+/// Generates a [MasonBundle] from the provided [brickPath] directory.
+MasonBundle createBundle(String brickPath) {
```
Comment (felangel) [inline packages/mason/lib/src/bundler.dart:46]: Is this change necessary? it's a breaking change and it doesn't appear to be necessary can we revert it?

Comment (renancaraujo) [reply packages/mason/lib/src/bundler.dart]: Removing it 🗡️

File: packages/mason_cli/lib/src/commands/bundle.dart (hunk 2 above)
Comment (renancaraujo) [reply packages/mason_cli/lib/src/commands/bundle.dart:89]: As the `add` command doesn't allow specific versions, I didn't want to overstep the implementation here.

File: packages/mason_cli/lib/src/commands/bundle.dart (hunk 1 above)
Comment (felangel) [reply packages/mason_cli/lib/src/commands/bundle.dart:102]: This will have the side-effect of making the bricks visible in `mason ls -g` not sure if that's desirable or not.

Comment (renancaraujo) [reply packages/mason_cli/lib/src/commands/bundle.dart]: Not desirable, changed to use systemp.

File: packages/mason/test/src/mason_yaml_test.dart [hunk 6]
```diff
@@ -76,6 +76,15 @@ void main() {
             .having((g) => g.path, 'path', instance.git!.path),
       );
     });
+
+    test('isLocal', () {
+      // ignore: avoid_redundant_argument_values
+      final remoteInstance = BrickLocation(version: 'any');
```
Comment (felangel) [inline packages/mason/test/src/mason_yaml_test.dart:82]: minor but can we add a check for a git based brick location?

File: packages/mason_cli/lib/src/commands/bundle.dart [hunk 7]
```diff
@@ -81,6 +135,7 @@ class BundleCommand extends MasonCommand {
       rethrow;
     }
 
+    tempBricksJson.clear();
```
Comment (felangel) [inline packages/mason_cli/lib/src/commands/bundle.dart:138]: 💯

File: packages/mason/test/src/mason_yaml_test.dart (hunk 6 above)
Comment (renancaraujo) [reply packages/mason/test/src/mason_yaml_test.dart:82]: Totally forgot that, thanx

**Review by felangel — APPROVED** [review]
LGTM 🎉 
Thanks so much for the contribution! 💙
